File: apps/api/app/schemas/meta.py
"""Pydantic models shared by metadata search endpoints."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field

MetaItemType = Literal["movie", "tv", "album"]


class MetaSearchItem(BaseModel):
    type: MetaItemType
    provider: str
    id: str
    title: str
    subtitle: str | None = None
    year: int | None = None
    poster: str | None = None
    extra: dict[str, Any] | None = None


class MetaSearchResponse(BaseModel):
    items: list[MetaSearchItem] = Field(default_factory=list)


class MetaCastMember(BaseModel):
    name: str
    character: str | None = None


class MetaTVInfo(BaseModel):
    seasons: int | None = Field(default=None, ge=0)
    episodes: int | None = Field(default=None, ge=0)


class MetaTrack(BaseModel):
    position: str | None = None
    title: str
    duration: str | None = None


class MetaAlbumInfo(BaseModel):
    artist: str
    album: str
    year: int | None = None
    styles: list[str] = Field(default_factory=list)
    tracklist: list[MetaTrack] = Field(default_factory=list)


class CanonicalMovie(BaseModel):
    title: str
    year: int | None = None


class CanonicalTV(BaseModel):
    title: str
    season: int | None = Field(default=None, ge=0)
    episode: int | None = Field(default=None, ge=0)


class CanonicalAlbum(BaseModel):
    artist: str
    album: str
    year: int | None = None


class CanonicalPayload(BaseModel):
    query: str
    movie: CanonicalMovie | None = None
    tv: CanonicalTV | None = None
    album: CanonicalAlbum | None = None


class MetaDetail(BaseModel):
    type: MetaItemType
    title: str
    year: int | None = None
    poster: str | None = None
    backdrop: str | None = None
    synopsis: str | None = None
    genres: list[str] = Field(default_factory=list)
    runtime: int | None = Field(default=None, ge=0)
    rating: float | None = Field(default=None, ge=0.0)
    cast: list[MetaCastMember] = Field(default_factory=list)
    tv: MetaTVInfo | None = None
    album: MetaAlbumInfo | None = None
    canonical: CanonicalPayload


class StartIndexingPayload(BaseModel):
    type: MetaItemType
    canonicalTitle: str
    movie: CanonicalMovie | None = None
    tv: CanonicalTV | None = None
    album: CanonicalAlbum | None = None


__all__ = [
    "MetaItemType",
    "MetaSearchItem",
    "MetaSearchResponse",
    "MetaDetail",
    "MetaCastMember",
    "MetaTVInfo",
    "MetaAlbumInfo",
    "MetaTrack",
    "CanonicalMovie",
    "CanonicalTV",
    "CanonicalAlbum",
    "CanonicalPayload",
    "StartIndexingPayload",
]

File: apps/web/src/app/types/meta.ts
export type MetaItemType = 'movie' | 'tv' | 'album';

export interface MetaSearchItem {
  type: MetaItemType;
  provider: string;
  id: string;
  title: string;
  subtitle?: string;
  year?: number;
  poster?: string;
  extra?: Record<string, unknown>;
}

export interface MetaSearchResponse {
  items: MetaSearchItem[];
}

export interface MetaCastMember {
  name: string;
  character?: string | null;
}

export interface MetaTVInfo {
  seasons?: number | null;
  episodes?: number | null;
}

export interface MetaTrack {
  position?: string | null;
  title: string;
  duration?: string | null;
}

export interface MetaAlbumInfo {
  artist: string;
  album: string;
  year?: number | null;
  styles: string[];
  tracklist: MetaTrack[];
}

export interface CanonicalMovie {
  title: string;
  year?: number | null;
}

export interface CanonicalTV {
  title: string;
  season?: number | null;
  episode?: number | null;
}

export interface CanonicalAlbum {
  artist: string;
  album: string;
  year?: number | null;
}

export interface CanonicalPayload {
  query: string;
  movie?: CanonicalMovie | null;
  tv?: CanonicalTV | null;
  album?: CanonicalAlbum | null;
}

export interface MetaDetail {
  type: MetaItemType;
  title: string;
  year?: number | null;
  poster?: string | null;
  backdrop?: string | null;
  synopsis?: string | null;
  genres: string[];
  runtime?: number | null;
  rating?: number | null;
  cast: MetaCastMember[];
  tv?: MetaTVInfo | null;
  album?: MetaAlbumInfo | null;
  canonical: CanonicalPayload;
}

